### DISCUSSION TITLE
DuckDB instead of SQLite?

### DISCUSSION THREAD

**Discussion by tomtom215** [mmcc-xx/BirdCAGE#23] [Ideas]
I have been running the BirdNetPi project for over a year. I have over 100k detections and some days I get over 1000 detections alone. 

Now that I have a large dataset, any of the analysis (species by day/month, etc) is very slow. Since this project may see the same scale I wanted to put out the idea of leveraging DuckDB which may be better suited for that type of analytics but is as light as SQLite: https://duckdb.org/

**Comment by mmcc-xx** [comment]
I'll check it out - thanks for the recommendation.

**Comment by mmcc-xx** [comment]
Can you tell me what kinds of reports you like to have available? I'd like to add some reporting pretty soon.

**Reply by matthew73210** [reply]
Maybe something like this as an easy option : https://github.com/ydataai/ydata-profiling and an robust bd connection for juypter ;)

**Comment by mmcc-xx** [comment]
So I've just added a weekly and monthly report, but I think what you have in mind is something broader. What would this application need to provide in order to play nice with ydata-profiling and Jupyter? I have only messed around with jupyter minimally, and I have no idea what pandas are other than being cute Chinese bears.

**Reply by matthew73210** [reply]
Hey,
Sorry for the late reply. Pandas is a lib on python that allows you to manipulate and analyse large datasets. And juypter is just a front end to a python env. All you really need is an API where you can grab a csv with all the data from the db. Like the whole set or between certain values. After that the onus is on the user to manipulate the data how he wants :)
I suppose you could run a python script that grabs the data from the db, ingest it in a Pandas dataframe and then passes it on to ydataai, which generates a HTML and passes it on to the front end to be displayed. The quickstart idea seems like a good place to start.
First of all, is there a simple API where the whole db can be grabbed in csv format? I haven't checked tbh

Cheers

**Reply by mmcc-xx** [reply]
There is not, but it would be easy enough to implement.

Just out of curiosity, what sorts of "questions" would you like to ask of the data?

**Reply by tomtom215** [reply]
> Hey,
> 
> Sorry for the late reply. Pandas is a lib on python that allows you to manipulate and analyse large datasets. And juypter is just a front end to a python env. All you really need is an API where you can grab a csv with all the data from the db. Like the whole set or between certain values. After that the onus is on the user to manipulate the data how he wants :)
> 
> I suppose you could run a python script that grabs the data from the db, ingest it in a Pandas dataframe and then passes it on to ydataai, which generates a HTML and passes it on to the front end to be displayed. The quickstart idea seems like a good place to start.
> 
> First of all, is there a simple API where the whole db can be grabbed in csv format? I haven't checked tbh
> 
> 
> 
> Cheers

At this point wouldn't it be better to use Polars instead of pandas? And also Streamlit instead of Jupyter notebooks?

**Reply by matthew73210** [reply]
Not wrong. https://www.datarevenue.com/en-blog/data-dashboarding-streamlit-vs-dash-vs-shiny-vs-voila sums it up pretty well. I use Jupyter and prefer it, but if you have a robust API it doesn't really matter. I just liked the auto HTML report that came with  https://github.com/ydataai/ydata-profiling, but you can do that with anything tbh.

**Reply by matthew73210** [reply]
> There is not, but it would be easy enough to implement.
> 
> 
> 
> Just out of curiosity, what sorts of "questions" would you like to ask of the data?

Ah didn't see your question, sorry. Generally looking for correlations between different birds, time of day, weather, food outside, cats etc. And data manipulation to make some nice graphs 
I'm gonna see what I can do with a db export and play around with the data. If it works I'll have a go at building birdcage locally (build not use the images) and adding the report through a PR.

**Reply by matthew73210** [reply]
Carring on, i did an extraction with sqlite web. Not very interesting at the moment, but probs lacking data. How feasible would it be to add weather reporting? Nothing too fancy, something like temp and conditions per hour.

Edit :  Just had a quick go with the python package request facing https://wttr.in/lat,long and it works a treat. https://wttr.in/:help sums all the options. I'll see what I can do.

**Comment by matthew73210** [comment]
Any chance, when you have the time. An option to export the .db as a csv (maybe in each report area) and also add an indication to the size of the current db + recordings ?
